### PR TITLE
Added logic handling near-clifford and noisy circuits. Removed errone…

### PR DESCRIPTION
**Pull request by asciineuron**
Allows a single state evolution to track all bitstring samples simultaneously, unless near-clifford or non-unitary (e.g. noisy) circuit, in which we fall back to previous behavior

### CODE REVIEW

File: bgls/simulator_test.py [hunk 1]
```diff
@@ -210,40 +210,6 @@ def test_run_with_stabilizer_ch_simulator():
     assert result_stabilizer_ch == result_state_vector
 
 
-def test_run_with_stabilizer_ch_simulator_near_clifford():
```
Comment (rmlarose) [inline bgls/simulator_test.py:213]: Why is this removed?

Comment (asciineuron) [reply bgls/simulator_test.py:213]: Digging into this more I am a bit confused. Shouldn't we expect the sample outcomes to actually differ somewhat since act_on_near_clifford only approximates the T gates? If I run it on wider randomized circuits the results don't necessarily match.

Comment (rmlarose) [reply bgls/simulator_test.py:213]: Given non-Clifford `U` the expansion  is `U = \sum_{i = 1}^{N} a_i C_i` where `a_i` are real and `C_i` are Clifford, correct? I believe this is equality if you keep all `N` terms. If the expansion is approximate or if you keep less than `N` terms then there's of course approximation error.

But in any case comparing sampled bitstrings is ill defined in this setting. A better test is to check that the expectation value of a set of observables is (approximately) equal.

File: bgls/simulator.py [hunk 2]
```diff
@@ -201,3 +210,23 @@ def _sample_bitstrings(
                     )
                 ]
         return keys_to_bitstrings
+
+    def _needs_trajectories(self, circuit: "cirq.AbstractCircuit") -> bool:
+        """Determines if repeated samples can be drawn for a single
+        simulation. For near-clifford, noisy, or non-unitary circuits this
+        is not possible. See qsim/qsimcirq/qsim_simulator.py"""
```
Comment (rmlarose) [inline bgls/simulator.py:217]: "See qsim/qsimcirq/qsim_simulator.py" is enough, but "Taken from https://github.com/quantumlib/qsim/blob/235ae2fc039fb4a98beb4a6114d10c7f8d2070f7/qsimcirq/qsim_simulator.py#L29" is also an option. As in qsim this is really a static method and could be outside the class.

**Review by rmlarose — CHANGES_REQUESTED** [review]
The fact that `_sample` and `_sample_bitstrings` have the same signature (except for the rtype) and the same docstring and essentially the same name is super confusing. At minimum the difference should be explained in the docstrings, and ideally in the docstrings and method names.

File: bgls/simulator.py (hunk 2 above)
Comment (asciineuron) [reply bgls/simulator.py]: Got it yeah I just added apply_gate as a parameter and was able to move it outside the class

File: bgls/simulator_test.py [hunk 3]
```diff
@@ -278,3 +246,53 @@ def test_remains_clifford():
...
+    """
+    qs = cirq.LineQubit.range(3)
+    circuit = bgls.utils.generate_random_circuit(
+        qs,
+        n_moments=10,
+        op_density=0.5,
+        gate_domain={cirq.S, cirq.CNOT, cirq.H, cirq.T},
+        random_state=1,
+    )
+    circuit.append(cirq.measure(qs, key="meas"))
```
[the remark this answers is not in the document]
Comment (asciineuron) [reply bgls/simulator_test.py]: Just removed this - I had just been habituated to adding it since usually I'm looking at histograms, but no reason for it here

**Comment by asciineuron** [comment]
I updated the confusing function naming, hopefully should be clearer now! The _sample docstring just comes from cirq's corresponding function, so I tried to make the other function more descriptive.

**Review by rmlarose — APPROVED** [review]
LGTM and can merge after resolving conflicts
